### ISSUE TITLE
Amazon S3 Tables Integration

### ISSUE THREAD

**Issue by flyrain** [apache/polaris#577]
### Describe the solution you'd like

Polaris is designed to act as a **REST facade for S3 tables**, enabling both **read** and **write** operations by interacting with the S3 table API. Polaris registers an S3 table using its **metadata location**. A flag will be needed to label the new Iceberg table is a s3 table. Below is a summary of the proposed approach:
### **Read Path**  
- The **LoadTable** endpoint in Polaris will call the S3 table API `get_table_metadata_location` to fetch the `metadata.json` file.  
- Polaris will then **serialize** the content of the `metadata.json` into a **LoadTableResponse** to return to the client.

### **Write Path**  
When the table is updated, Polaris will:  
1. Gather the changes and generate a new `metadata.json`.  
2. Use the S3 table API [`update-table-metadata-location`](https://docs.aws.amazon.com/cli/latest/reference/s3tables/update-table-metadata-location.html) to commit the new metadata.  

### **AuthZ/AuthN**
We need to ensure that the AWS role used for creating the Polaris catalog has the read and write privileges of the s3 table.

**Comment by puchengy** [comment]
@flyrain Thank you for starting the issue, I have two questions:
1. is it right that with this integration, users will continue to have the option to own the metastore service (that holds source of the truth data)?
2. should it be a concern if "update-table-metadata-location" fail and cause table metadata out of the sync, which could mislead S3 Tables to clean up data as orphan files.

**Comment by flyrain** [comment]
> is it right that with this integration, users will continue to have the option to own the metastore service (that holds source of the truth data)?

That's right. This integration won't change the source of truth(s3 table in this case), and other tools or pipelines against the source catalog should still work as is.

> should it be a concern if "update-table-metadata-location" fail and cause table metadata out of the sync, which could mislead S3 Tables to clean up missing metadata data (and its data files) as orphan files?

Polaris client will get the failure message in that case, then it can retry or just fail itself. The table is still consistent. But the failure may leave orphan files, which is fine as the other clients also leave orphan files in case of failure. 

Thanks for pointing out the wrong link. Updated.

**Comment by puchengy** [comment]
@flyrain 

> Polaris client will get the failure message in that case, then it can retry or just fail itself. The table is still consistent.

Do you mean that "update-table-metadata-location" will be part of the commit to the source catalog (e.g. HMS)?

**Comment by flyrain** [comment]
Yes, you could consider Polaris as a proxy when the source of truth is a remote catalog.  Here are two different commit paths depends on different client types:
```
# REST clients commit path
REST client ->  Polaris -> Remote catalog(e.g. HMS, S3 table)

# Other clients commit path
Non-REST clients ->  Remote catalog(e.g. HMS, S3 table)
```

**Comment by puchengy** [comment]
@flyrain Got it, so in this proposal, HMS and S3 Tables are mutually exclusive? Sorry, this was my real intention of asking that question 

> is it right that with this integration, users will continue to have the option to own the metastore service (that holds source of the truth data)?

My major goal is to see if it is possible to operate HMS and S3 Tables at the same time, while keeping tables in HMS (because we want to keep our data in our service).

**Comment by flyrain** [comment]
Yes. They are mutually exclusive. I think it's better to only have one source of truth. In case of s3 tables, s3 service is the source of the truth for sure. An integration of HMS would be like:
```
HMS clients ->  HMS -> S3 table
```
HMS will have to invoke the s3 api `update-table-metadata-location` to commit, otherwise there would be consistent issue.

It's another topic anyway. We may discuss it elsewhere.

**Comment by soumilshah1995** [comment]
I wanted to kindly check if there are any updates on when these items are expected to be released

**Comment by donPain** [comment]
Is it currently possible to create a federated catalog for S3Tables in Polaris?

**Comment by flyrain** [comment]
Polaris federation supports three types of remote catalog, REST, HIVE, and HADOOP. Given that you can visit s3 table via IRC now[1], the answer is Yes.

1. https://docs.aws.amazon.com/AmazonS3/latest/userguide/s3-tables-integrating-glue-endpoint.html
